arangod/Aql/Functions.cpp

```diff
@@ -8698,6 +8698,178 @@ AqlValue Functions::CallGreenspun(arangodb::aql::ExpressionContext* expressionCo
   }
 }
 
+static VPackBuilder buildKeyObject(VPackStringRef key, bool closeObject = true) {
+  VPackBuilder builder;
+  builder.openObject(true);
+  builder.add(StaticStrings::KeyString, VPackValuePair(key.data(), key.size(), VPackValueType::String));
+  if (closeObject) {
+    builder.close();
+  }
+  return builder;
+}
+
+static AqlValue ConvertToObject(VPackSlice input, bool allowKeyConversionToObject, bool canUseCustomKey, bool ignoreErrors) {
+  // input is not an object.
+  // if this happens, it must be a string key
+  if (!input.isString() || !allowKeyConversionToObject) {
+    if (ignoreErrors) {
+      return AqlValue{};
+    }
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_ARANGO_DOCUMENT_TYPE_INVALID);
+  }
+
+  // check if custom keys are allowed in this context
+  if (!canUseCustomKey) {
+    if (ignoreErrors) {
+      return AqlValue{};
+    }
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_CLUSTER_MUST_NOT_SPECIFY_KEY);
+  }
+
+  // convert string key into object with { _key: "string" }
+  TRI_ASSERT(allowKeyConversionToObject);
+  return AqlValue{buildKeyObject(input.stringRef()).slice()};
```

Comment:
This will call `buildKeyObject(...)` for each input, which creates a VPackBuilder temporary.
This VPackBuilder object will be thrown away directly afterwards.

Can we pass in a temporary Builder object into `ConvertToObject`, so an existing Builder can be recycled? Alternatively, we can pass in the current transaction and use sth like
```
transaction::BuilderLeaser tempBuilder(trx);
```

Reply:
Ah, I did not know about `BuilderLeaser`, but that makes sense!